website/blog/en/2020/pixel-bencmhark.md

```diff
@@ -0,0 +1,85 @@
+
+```
+Thread 60 "ConfigReloader" received signal SIGSEGV, Segmentation fault.                         
+[Switching to LWP 21873]                        
+0x000000556a73f740 in ?? ()          
+
+(gdb) whe 20                                    
+#0  0x000000556a73f740 in ?? ()                 
+#1  0x000000556a744028 in ?? ()                 
+#2  0x000000556a73e5a0 in ?? ()                 
+#3  0x000000556a73d250 in unw_init_local ()     
+#4  0x000000556a73deb8 in unw_backtrace ()      
+#5  0x0000005562aabb54 in StackTrace::tryCapture() ()                                           
+#6  0x0000005562aabb10 in StackTrace::StackTrace() ()                                           
+#7  0x0000005562a8d73c in MemoryTracker::alloc(long) ()                                         
+#8  0x0000005562a8db38 in MemoryTracker::alloc(long) ()                                         
+#9  0x0000005562a8e8bc in CurrentMemoryTracker::alloc(long) ()                                  
+#10 0x0000005562a8b88c in operator new[](unsigned long) ()                                      
+#11 0x0000005569c35f08 in Poco::XML::NamePool::NamePool(unsigned long) ()                       
+...
+```
+
+What is this function, and why do we need it? In this particular stack trace, we're out of memory, and about to throw an exception for that. `unw_backtrace` is called to produce a backtrace for the exception message. But there is another interesting context where we call it. Believe it or not, ClickHouse has a built-in `perf`-like sampling profiler that can save stack traces for CPU time and real time, and also memory allocations. The data is saved into a `system.trace_log` table, so you can build flame graphs for what your query was doing as simple as piping output of an SQL query into `flamegraph.pl`. This is an interesting feature, but what is relevant now is that it sends signals to all threads of the server to interrupt them at some random time and save their current backtraces, using the same `unw_backtrace` function that we know to segfault. We expect query profiler to be used in production environment, so it is enabled by default. After disabling it, we have a functioning ClickHouse server running on Android.
```

Comment:
```suggestion
What is this function, and why do we need it? In this particular stack trace, we're out of memory, and about to throw an exception for that. `unw_backtrace` is called to produce a backtrace for the exception message. But there is another interesting context where we call it. Believe it or not, ClickHouse has a built-in `perf`-like sampling profiler that can save stack traces for CPU time and real-time, and also memory allocations. The data is saved into a `system.trace_log` table, so you can build flame graphs for what your query was doing as simple as piping output of an SQL query into `flamegraph.pl`. This is an interesting feature, but what is relevant now is that it sends signals to all threads of the server to interrupt them at some random time and save their current backtraces, using the same `unw_backtrace` function that we know to segfault. We expect query profiler to be used in a production environment, so it is enabled by default. After disabling it, we have a functioning ClickHouse server running on Android.
```

Reply:
Real time is not an adjective, it's a noun + modifier. Adding a dash between them is wrong.